### ISSUE TITLE
[FeatureRequest] Fetch updates from GitHub Releases page

### ISSUE THREAD

**Issue by k06a** [sparkle-project/Sparkle#648]
My app binaries are hosted on GitHub Releases page.
I just wanna Sparkle to check this page for new versions :)

**Comment by zorgiepoo** [comment]
It may be more practical to generate an appcast server-side based on changes to GitHub's release page. I don't really know how much info GitHub gives you in order to do that in a sane manner however.

**Comment by kainjow** [comment]
You could subscribe to the Atom page and update your Sparkle version once a new release comes out. Probably the safest way.

https://github.com/sparkle-project/Sparkle/releases.atom

**Comment by zorgiepoo** [comment]
Briefly looking at the atom feed, it's not meant to be used for this kind of task. Probably you'd want [GitHub's Releases API](https://developer.github.com/v3/repos/releases/).

**Comment by kainjow** [comment]
Well I meant doing it manually, no automation, since you would want to probably wait for a release to update in case there are bugs, especially if you're about to ship your software.

**Comment by zorgiepoo** [comment]
Don't see the point of subscribing to the feed if you are the one pushing the release onto GitHub and have go generate the appcast manually. Anycase, I thought the original poster wanted it to be automated. One could try to make use of the "pre-release" field GitHub has, too.

**Comment by k06a** [comment]
Really I would like to setup update suggestion for my app without any own backend. Don't wanna care of this dependency never. Just wanna add binaries to GitHub Releases page.

**Comment by pornel** [comment]
That sounds like a very convenient thing indeed.

I do have some concerns:

* Supporting GitHub's releases page directly would require Sparkle using GitHub API directly. That's a very specific dependency for Sparkle, and although GitHub is great and all, it is a dependency on a non-standard API of a closed-source proprietary product.

* Without an appcast there are no DSA signatures available, so we'd need to figure out what level of security we want, and how to implement it.


How about a slightly different thing:

a tool that generates appcast.xml file from a GitHub's releases page? (and maybe even a local directory containing binaries)

**Comment by k06a** [comment]
@pornel do you mean include `appcast.xml` in git repository files? I think it can be done, by filling template and appending to xml file or add terminal tool to add versions to appcast.xml.

**Comment by uranusjr** [comment]
While most information can be built automatically, the GitHub release page do lack some information essential for the appcast file, e.g. the version string (that positive incremental integer thing). Some extra metadata would still be needed.

**Comment by pornel** [comment]
@k06a Yes, something like that. Committed to the repo, or perhaps to gh-pages?

So it may be possible to do that with a tool that:

# Assumes it's run in a .git clone from github, so `git remote -v` will let you guess URL of the github releases page
# Finds the releases
# Downloads the releases (?) to sign them with a private DSA key
# Generates appcast.xml
# Commits appcast.xml to gh-pages of the project (which must match appcast URL in Info.plist)

So that would be possible to do as a standalone tool outside of Sparkle.

**Comment by k06a** [comment]
Maybe called like this:
```
ruby release_appcast.rb appcast.xml https://github.com/sparkle-project/Sparkle/releases/download/1.12.0a2/Sparkle-1.12.0a2.tar.bz2
git commit -m "Released new version"
git push
```
Or `git` commands can be included inside script.

**Comment by huangyq23** [comment]
I have a similar attempt recently.

With some scripting and naming conventions. it is possible to encode the version number into the binary archive's filename.

I had a script to perform archive and export the app zip file with names like `MyApp.v1.2.b20151011.zip`.
Here, `1.2` and `20151011` maps to `CFBundleShortVersionString` and `CFBundleVersion` extracted from `Info.plist`.

With that, I can draft up and publish the releases and upload the binary just as usual in github web interface.

Then there is the matter of setting up the gh-pages correctly to generate the `appcast.xml`. Since github already have all the [releases objects|https://help.github.com/articles/repository-metadata-on-github-pages/] supplied into the rendering context as site metadata, it is possible to build the entire xml within the Jekyll process using [Liquid](http://jekyllrb.com/docs/templates/).

I filter out all draft and pre-releases for my `appcast.xml`. I also have a separate `appcast_pre.xml` that includes pre-releases for my alpha/beta channel.

The biggest benefit of render the xml in Github is no extra interaction with the Github Release API is required; normal git access will suffice. Obviously this is strong convention and less flexible since the resolving logic relies on the filename, but it suits my objectives well.

I can share my build script and liquid template for `appcast.xml` if anyone is interested.

**Comment by k06a** [comment]
@huangyq23 will be perfect to share you result with community! 👍

**Comment by huangyq23** [comment]
@k06a 
Just got some time to work on this. I have them uploaded to a [gist](https://gist.github.com/huangyq23/668e6d6fcccf714e802a).

I also wrote a [blogpost](http://yiqiu.me/programming/2015/11/19/sparkle-update-on-github.html) explaining some of the thought behind the solution.

Hope it helps.

**Comment by pornel** [comment]
@huangyq23 That's great! Thank you very much

**Comment by orta** [comment]
The [CocoaPods app](https://github.com/cocoapods/cocoapods-app) uses the same kind of technique as @huangyq23 mentioned, incase you want more examples. See the bottom of the [Rakefile](https://github.com/CocoaPods/CocoaPods-app/blob/master/Rakefile)

**Comment by pierredavidbelanger** [comment]
As a lazy person, I do not want to have to maintain `appcast.xml` after each release.
So, I just made [a project to solve this exact problem](https://github.com/pierredavidbelanger/githubappcaster) for myself.
Then I realised it may help someone else on this thread.

**Comment by lwouis** [comment]
If someone is interested in a lightweight solution that doesn't require a backend (i.e. only using github to host the releases and `appcast.xml`), you can take a look at [my approach here](https://github.com/lwouis/alt-tab-macos/blob/v3-poc/scripts/update_appcast.sh).

**Comment by noah-nuebling** [comment]
See my solution [here](https://github.com/noah-nuebling/mac-mouse-fix/tree/update-feed)

It's a little verbose and very slow, but it has these perks

- It doesn't require a backend
- It creates a separate appcast for prereleases
- It automatically generates update notes. They look nice and native and support darkmode.
- It's written in python and I documented it pretty well, so you should be able to adapt it to your needs without too much trouble

Thanks @lwouis for the inspiration!

**Comment by zorgiepoo** [comment]
While I do appreciate all the workflows that have been contributed here, I think with regards to this issue still being open here, I'd prefer to leverage off generate_appcast which officially supports Sparkle's full feature set and numerous edge cases going forward (including delta updates, among other things). That might involve a adding an option to select a specific version to upload a draft (pre)release to a Github Pages repository.

**Comment by mikehearn** [comment]
For those who need this feature today, [Conveyor](https://www.hydraulic.dev) is a tool that makes Sparkle-ized Mac apps (and Windows, and Linux packages) from cross-platform app frameworks like Electron/Flutter/JVM or indeed native apps.

It can publish directly to GitHub Releases, also uploading a simple download page to GitHub Pages. And it can build/sign/notarize apps for all supported platforms from any platform, so you can run it in cheap GitHub CI workers or from your local dev laptop if you wish, you don't need a Mac to release Sparkleized Mac apps.

It's commercial (disclosure: by my company), but it's free for open source apps and can be used today to release to GH Releases.

**Comment by aehlke** [comment]
@zorgiepoo is there any tool available that uses that yet?

**Comment by chigkim** [comment]
I'd also love to see this feature!

**Comment by Usiel** [comment]
I came here as well looking for an easy way to get an appcast.xml up and running. I ended up building a basic WordPress plugin that can host and generate the appcast.xml, [here it is](https://wordpress.org/plugins/sparkling-appcast/) in case anyone is interested.
